lightning/src/ln/blinded_payment_tests.rs

```diff
@@ -1416,6 +1430,65 @@ fn custom_tlvs_to_blinded_path() {
 	);
 }
 
+#[test]
+fn fails_receive_tlvs_authentication() {
+	let chanmon_cfgs = create_chanmon_cfgs(2);
+	let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+	let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+	let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+	let chan_upd = create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 1_000_000, 0).0.contents;
+
+	let amt_msat = 5000;
+	let (_, payment_hash, payment_secret) = get_payment_preimage_hash(&nodes[1], Some(amt_msat), None);
+	let payee_tlvs = UnauthenticatedReceiveTlvs {
+		payment_secret,
+		payment_constraints: PaymentConstraints {
+			max_cltv_expiry: u32::max_value(),
+			htlc_minimum_msat: chan_upd.htlc_minimum_msat,
+		},
+		payment_context: PaymentContext::Bolt12Refund(Bolt12RefundContext {}),
+	};
+	let nonce = Nonce([42u8; 16]);
+	let expanded_key = chanmon_cfgs[1].keys_manager.get_inbound_payment_key();
+	let mut payee_tlvs = payee_tlvs.authenticate(nonce, &expanded_key);
+	// Swap in a different nonce to force authentication to fail.
+	payee_tlvs.authentication.1 = Nonce([0u8; 16]);
+
+	let mut secp_ctx = Secp256k1::new();
+	let blinded_path = BlindedPaymentPath::new(
+		&[], nodes[1].node.get_our_node_id(), payee_tlvs, u64::MAX, TEST_FINAL_CLTV as u16,
+		&chanmon_cfgs[1].keys_manager, &secp_ctx
+	).unwrap();
+
+	let route_params = RouteParameters::from_payment_params_and_value(
+		PaymentParameters::blinded(vec![blinded_path]),
+		amt_msat,
+	);
+
+	nodes[0].node.send_payment(payment_hash, RecipientOnionFields::spontaneous_empty(), PaymentId(payment_hash.0), route_params, Retry::Attempts(0)).unwrap();
+	check_added_monitors(&nodes[0], 1);
+
+	let mut events = nodes[0].node.get_and_clear_pending_msg_events();
+	assert_eq!(events.len(), 1);
+	let ev = remove_first_msg_event_to_node(&nodes[1].node.get_our_node_id(), &mut events);
+	let mut payment_event = SendEvent::from_event(ev);
+
+	nodes[1].node.handle_update_add_htlc(nodes[0].node.get_our_node_id(), &payment_event.msgs[0]);
+	check_added_monitors!(nodes[1], 0);
+	do_commitment_signed_dance(&nodes[1], &nodes[0], &payment_event.commitment_msg, true, true);
+	nodes[1].node.process_pending_htlc_forwards();
```

Comment:
Oh, btw I think you might be able to replace this with something like:
```rust
	nodes[0].node.send_payment(payment_hash, RecipientOnionFields::spontaneous_empty(),
	PaymentId(payment_hash.0), route_params, Retry::Attempts(0)).unwrap();
	check_added_monitors(&nodes[0], 1);
	pass_along_route(&nodes[0], &[&[&nodes[1]]], amt_msat, payment_hash, payment_secret);
```
Taken from `do_one_hop_blinded_path`. But not a big deal.

Reply:
Hmmm... I see the following failure using that since the HTLC should fail, IIUC.

```
thread 'ln::blinded_payment_tests::fails_receive_tlvs_authentication' panicked at 'assertion failed: node_a.node.get_and_clear_pending_msg_events().is_empty()', lightning/src/ln/functional_test_utils.rs:2141:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:48:5
   3: lightning::ln::functional_test_utils::do_commitment_signed_dance
             at ./src/ln/functional_test_utils.rs:2141:3
   4: lightning::ln::functional_test_utils::do_pass_along_path
             at ./src/ln/functional_test_utils.rs:2710:4
   5: lightning::ln::functional_test_utils::pass_along_path
             at ./src/ln/functional_test_utils.rs:2789:2
   6: lightning::ln::functional_test_utils::pass_along_route
             at ./src/ln/functional_test_utils.rs:2820:3
   7: lightning::ln::blinded_payment_tests::fails_receive_tlvs_authentication
             at ./src/ln/blinded_payment_tests.rs:1470:2
   8: lightning::ln::blinded_payment_tests::fails_receive_tlvs_authentication::{{closure}}
             at ./src/ln/blinded_payment_tests.rs:1434:1
   9: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
  10: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Reply:
Oh, right, the mentioned test method does include a failure but it actively fails the payment. Not gonna bother looking into it further lol.